### PR TITLE
change(state): Add state requests and support code for the `z_getsubtreesbyindex` RPC

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

This PR contains the state requests and `zebra-chain` support code used by the `z_getsubtreesbyindex` RPC.

It can merge before or after the release, because it doesn't add any new RPC methods or change finalized state data.

### Specifications

See https://github.com/zcash/zcash/pull/6677/files#diff-decae4be02fb8a47ab4557fe74a9cb853bdfa3ec0fa1b515c0a1e5de91f4ad0bR1458-R1475

The specs for this RPC aren't on the RPC spec site yet.

### Complex Code or Requirements

When combining subtrees from the finalized and non-finalized states, we need to deal with an overlap between the finalized state, and a non-finalized chain that is now a side-chain. This is rare, but it can happen. See the code comments for details.

## Solution

[Define state request and response types for subtrees](https://github.com/ZcashFoundation/zebra/pull/7408/commits/9887602d455146c11f19be179b3158f8d881e5bb)
   - Remove `Arc` wrappers from subtree types, it's only 38 `Copy`-able bytes

[Implement consistent concurrent subtree read requests](https://github.com/ZcashFoundation/zebra/pull/7408/commits/af4219a23ad44871b59afbbde3db22f5b7103144)
   - [Add finalized state subtree list methods and delete unused methods](https://github.com/ZcashFoundation/zebra/pull/7408/commits/b30cebf14e30886ae95f6efa18898717a00e6df8)
   - [Implement subtree list lookups in the non-finalized state and delete unused methods](https://github.com/ZcashFoundation/zebra/pull/7408/commits/cd0baa759595bb944d6e8ab6520bad3c5aea8b11)
   - [Implement FromDisk for NoteCommitmentSubtreeIndex and add tests](https://github.com/ZcashFoundation/zebra/pull/7408/commits/740d4edf7c1fdeb2c997ab6fc215495667a43bad)
 
Implement zebra-chain support for the RPC parameter and return types:
   - [Make NoteCommitmentSubtreeIndex compatible with serde-based RPCs](https://github.com/ZcashFoundation/zebra/pull/7408/commits/f48fc17cc3e220461fed4455f20318df77f7d171)
   - [Implement ToHex for sapling::Node](https://github.com/ZcashFoundation/zebra/pull/7408/commits/f156e4cc1722c1187214e04fd7f5609c62cc5def)
   - [Implement ToHex for orchard::Node](https://github.com/ZcashFoundation/zebra/pull/7408/commits/6e8a825b6c19520f3d7862d1f05fcb3a8fa6cd82)

### Testing

The code in this PR will be tested along with the RPC method, when that gets added in the next PR.

## Review

This PR can be reviewed and merged at any time.

It has a minor merge conflict with PR #7437, the added code and comments in both PRs should be kept.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

### CODE REVIEW

File: zebra-rpc/src/methods.rs
Comment (teor2345) [inline zebra-rpc/src/methods.rs]: This PR is ready for review, but we can't merge this new RPC until after the 1.2.0 release, because it doesn't have any data yet.

File: zebra-state/src/request.rs
Comment (upbqdn) [inline zebra-state/src/request.rs:859]: ###### Optional

We can make this more consistent with the rest of the requests:
```suggestion
    /// * An empty list if there is no subtree at `start_index`
    ///
```

File: zebra-state/src/request.rs
Comment (upbqdn) [inline zebra-state/src/request.rs:874]: ###### Optional

We can make this more consistent with the rest of the requests:
```suggestion
    /// * An empty list if there is no subtree at `start_index`
    ///
```

File: zebra-state/src/service/read/tree.rs
Comment (upbqdn) [inline zebra-state/src/service/read/tree.rs:143]: The last sentence doesn't mention if it's the finalized or non-finalized tip. We could move the comment on correctness from below to here.

File: zebra-state/src/service/read/tree.rs
Comment (upbqdn) [inline zebra-state/src/service/read/tree.rs:76]: ###### Optional

```suggestion
    // for tree roots to be equal once the leaves have diverged.)
    
```

File: zebra-state/src/service/read/tree.rs
Comment (upbqdn) [inline zebra-state/src/service/read/tree.rs:164]: ###### Optional

```suggestion
    // for tree roots to be equal once the leaves have diverged.)
    
```

File: zebra-rpc/src/methods.rs
Comment (teor2345) [reply zebra-rpc/src/methods.rs]: I split the RPC method into another PR so this PR could merge.

File: zebra-state/src/service/read/tree.rs
Comment (upbqdn) [inline zebra-state/src/service/read/tree.rs:55]: The last sentence doesn't mention if it's the finalized or non-finalized tip. We could move the comment on correctness from below to here.

File: zebra-state/src/service/read/tree.rs
Comment (upbqdn) [inline zebra-state/src/service/read/tree.rs:76]: I would promote this comment to the docs for the whole function so that it's clear what it returns if there are different subtrees in the provided `db` and `chain`.

File: zebra-state/src/service/read/tree.rs
Comment (upbqdn) [inline zebra-state/src/service/read/tree.rs:164]: I would promote this comment to the docs for the whole function so that it's clear what it returns if there are different subtrees in the provided `db` and `chain`.

**Review by upbqdn — APPROVED** [review]
utACK

File: zebra-state/src/request.rs
Comment (teor2345) [reply zebra-state/src/request.rs:859]: I think these are good suggestions, and it seems like you have a good idea how you'd like the documentation to look. Would you mind submitting a PR with these changes?

Or are you happy for me to accept all your suggestions and leave it at that?

Comment (upbqdn) [reply zebra-state/src/request.rs:859]: I'll submit a PR.

**Comment by teor2345** [comment]
@mergifyio refresh

**Pull request merged**
